drivers/clock_control/clock_control_nrf2_common.c

```diff
@@ -0,0 +1,158 @@
+ */
+
+#include "clock_control_nrf2_common.h"
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(clock_control_nrf2);
+
+#define FLAG_UPDATE_IN_PROGRESS BIT(FLAGS_COMMON_BITS - 1)
+#define FLAG_UPDATE_NEEDED      BIT(FLAGS_COMMON_BITS - 2)
+
+#define ONOFF_CNT_MAX (FLAGS_COMMON_BITS - 2)
+
+#define CONTAINER_OF_ITEM(ptr, idx, type, array) \
+	(type *)((char *)ptr - \
+		 (idx * sizeof(array[0])) - \
+		 offsetof(type, array[0]))
+
+/*
+ * Definition of `struct clock_config_generic`.
+ * Used to access `clock_config_*` structures in a common way.
+ */
+STRUCT_CLOCK_CONFIG(generic, ONOFF_CNT_MAX);
+
+static void update_config(struct clock_config_generic *cfg)
+{
+	atomic_val_t prev_flags = atomic_or(&cfg->flags, FLAG_UPDATE_NEEDED);
+
+	/* If the update work is already scheduled (FLAG_UPDATE_NEEDED was
+	 * set before the above OR operation) or is currently being executed,
+	 * it is not to be submitted again. In the latter case, it will be
+	 * submitted by clock_config_update_end().
+	 */
+	if (prev_flags & (FLAG_UPDATE_NEEDED | FLAG_UPDATE_IN_PROGRESS)) {
+		return;
+	}
+
+	k_work_submit(&cfg->work);
+}
+
+static void onoff_start_option(struct onoff_manager *mgr,
+			       onoff_notify_fn notify)
+{
+	struct clock_onoff *onoff =
+		CONTAINER_OF(mgr, struct clock_onoff, mgr);
+	struct clock_config_generic *cfg =
+		CONTAINER_OF_ITEM(onoff, onoff->idx,
+				  struct clock_config_generic, onoff);
+
+	onoff->notify = notify;
+
+	(void)atomic_or(&cfg->flags, BIT(onoff->idx));
+	update_config(cfg);
+}
+
+static void onoff_stop_option(struct onoff_manager *mgr,
+			      onoff_notify_fn notify)
+{
+	struct clock_onoff *onoff =
+		CONTAINER_OF(mgr, struct clock_onoff, mgr);
+	struct clock_config_generic *cfg =
+		CONTAINER_OF_ITEM(onoff, onoff->idx,
+				  struct clock_config_generic, onoff);
+
+	(void)atomic_and(&cfg->flags, ~BIT(onoff->idx));
+	update_config(cfg);
+
+	notify(mgr, 0);
+}
+
+int clock_config_init(void *clk_cfg, uint8_t onoff_cnt,
+		      k_work_handler_t update_work_handler)
+{
+	struct clock_config_generic *cfg = clk_cfg;
+
+	__ASSERT_NO_MSG(onoff_cnt <= ONOFF_CNT_MAX);
+
+	for (int i = 0; i < onoff_cnt; ++i) {
+		static const struct onoff_transitions transitions = {
+			.start = onoff_start_option,
+			.stop  = onoff_stop_option
+		};
+		int rc;
+
+		rc = onoff_manager_init(&cfg->onoff[i].mgr, &transitions);
+		if (rc < 0) {
+			return rc;
+		}
+
+		cfg->onoff[i].idx = (uint8_t)i;
+	}
+
+	cfg->onoff_cnt = onoff_cnt;
+
+	k_work_init(&cfg->work, update_work_handler);
+
+	return 0;
+}
+
+uint8_t clock_config_update_begin(struct k_work *work)
+{
+	struct clock_config_generic *cfg =
+		CONTAINER_OF(work, struct clock_config_generic, work);
+	uint32_t active_options;
+
+	(void)atomic_or(&cfg->flags, FLAG_UPDATE_IN_PROGRESS);
+	cfg->flags_snapshot = atomic_and(&cfg->flags, ~FLAG_UPDATE_NEEDED);
+
+	active_options = cfg->flags_snapshot & BIT_MASK(ONOFF_CNT_MAX);
+
+	return active_options ? (uint8_t)(31 - __builtin_clz(active_options))
```

Comment:
I would move `(uint8_t)(31 - __builtin_clz(active_options))` to a `static inline` function with a name indicating what it is intended to calculate. Right now I'm guessing what was your intention :)

Reply:
Done in new PR